sdk/core/azure-core/src/http/policy_retry.cpp

```diff
@@ -0,0 +1,113 @@
+  auto const responseHeadersEnd = responseHeaders.end();
+  auto header = responseHeadersEnd;
+  if (((header = responseHeaders.find("retry-after-ms")) != responseHeadersEnd)
+      || ((header = responseHeaders.find("x-ms-retry-after-ms")) != responseHeadersEnd))
+  {
+    // The headers above are in milliseconds.
+    return std::chrono::milliseconds(std::stoi(header->second));
+  }
+
+  if ((header = responseHeaders.find("Retry-After")) != responseHeadersEnd)
+  {
+    // This header is in seconds.
+    return std::chrono::seconds(std::stoi(header->second));
+
+    // TODO: to be accurate, the Retry-After header is EITHER seconds, or a DateTime. So we need to
+    // write a parser for that (and handle the case when parsing seconds fails).
+    // More info:
+    // * Retry-After header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Retry-After
+    // * HTTP Date format: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Date
+    // * Parsing the date: https://en.cppreference.com/w/cpp/locale/time_get
+    // * Get system datetime: https://en.cppreference.com/w/cpp/chrono/system_clock/now
+    // * Subtract datetimes to get duration:
+    // https://en.cppreference.com/w/cpp/chrono/time_point/operator_arith2
+  }
+
+  constexpr auto lastBit = std::numeric_limits<decltype(attempt)>::digits
+      - (std::numeric_limits<decltype(attempt)>::is_signed ? 1 : 0);
+
+  // Scale exponentially: 1 x RetryDelay on 1st attempt, 2x on 2nd, 4x on 3rd, 8x on 4th ... all the
+  // way up to std::numeric_limits<decltype(attempt)>::max() * RetryDelay.
+  auto exponentialRetryAfter = retryOptions.RetryDelay
+      * ((attempt <= (lastBit - 1)) ? (1 << attempt)
+                                    : std::numeric_limits<decltype(attempt)>::max());
+
+  // jitterFactor is a random double number in the range [0.8 .. 1.3)
+  auto jitterFactor = 0.8 + static_cast<double>(std::rand() / RAND_MAX) * 0.5;
+
+  // Multiply exponentialRetryAfter by jitterFactor
+  exponentialRetryAfter = decltype(
+      exponentialRetryAfter)(static_cast<decltype(exponentialRetryAfter)::rep>(
+      (std::chrono::duration<double, decltype(exponentialRetryAfter)::period>(exponentialRetryAfter)
+       * jitterFactor)
+          .count()));
+
+  return std::min(exponentialRetryAfter, retryOptions.MaxRetryDelay);
+}
+} // namespace
+
+std::unique_ptr<Response> RetryPolicy::Send(
+    Context& ctx,
+    Request& request,
+    NextHttpPolicy nextHttpPolicy) const
+{
+  for (decltype(m_retryOptions.MaxRetries) attempt = 1;; ++attempt)
+  {
+    request.StartRetry();
+
+    auto response = nextHttpPolicy.Send(ctx, request);
+
+    auto retryAfter = ShouldRetry(*response.get(), m_retryOptions, attempt);
+    // If we are out of retry attempts, if a response is non-retriable (or simply 200 OK, i.e
+    // doesn't need to be retried), then ShouldRetry returns some negative value.
+    if (retryAfter.count() < 0)
+    {
+      return response;
+    }
+
+    response.reset();
```

Comment:
I don't know what this does. Maybe the method name should change.
We do have to make sure that the response's BodyStream is destructed (closing the socket) if we're retrying due to an successful HTTP operation (as opposed to an exception via a context timeout or something else.

Also, this loop must check the context to see if it has canceled - in this case, do not retry.

Reply:
`.reset()` is the `unique_ptr`'s method. It releases the object that it holds. So to answer your question, that destroys the HttpResponse which should result in descruction of the BodyStream.

Regarding the context: see the comment just right below. I know that Victor has the code to ThrowIfCanceled() in one of his branches, I will insert it there, once it is in master.

Reply:
Should we call `->Rewind()` or `.reset()` here?

Reply:
Anton says above that .reset is a member of unique_ptr and so this can't be renamed.
However, Jinming is right that the request's BodyStream's Rewind method MUST be called just before each call to Send.

Reply:
`nextHttpPolicy.Send(ctx, request);`  creates the response, on line 92, on each loop iteration. So, in case we retry, I call `unique_ptr::reset()` which deletes the response object, and calls its destructor. Do i need to call the reqind? Because on the next loop iteration, if we retry, I am going to be given another newly allocated response.

Reply:
@antkmsft Okay, whatever works is fine with me.

Reply:
You never rewind a response stream (it will always throw an exception).
For each retry, you MUST rewind the request body stream.

Reply:
Done - we call Rewind() on request now.